hail/python/hail/vds/variant_dataset.py

```diff
@@ -115,3 +115,96 @@ def write(self, path, **kwargs):
     def checkpoint(self, path, **kwargs) -> 'VariantDataset':
         self.write(path, **kwargs)
         return read_vds(path)
+
+    def validate(self):
+        """Eagerly checks necessary representational properties of the VDS."""
+
+        rd = self.reference_data
+        vd = self.variant_data
+
+        def error(msg):
+            raise ValueError(f'VDS.validate: {msg}')
+
+        rd_row_key = rd.row_key.dtype
+        if (not isinstance(rd_row_key, hl.tstruct)
+                or len(rd_row_key) != 1
+                or not rd_row_key.fields[0] == 'locus'
+                or not isinstance(rd_row_key.types[0], hl.tlocus)):
+            error(f"expect reference data to have a single row key 'locus' of type locus, found {rd_row_key}")
+
+        vd_row_key = vd.row_key.dtype
+        if (not isinstance(vd_row_key, hl.tstruct)
+                or len(vd_row_key) != 2
+                or not vd_row_key.fields == ('locus', 'alleles')
+                or not isinstance(vd_row_key.types[0], hl.tlocus)
+                or vd_row_key.types[1] != hl.tarray(hl.tstr)):
+            error(
+                f"expect variant data to have a row key {{'locus': locus<rg>, alleles: array<str>}}, found {vd_row_key}")
```

Comment:
In many cases, a VDS will have row keys of just `locus`, it may be better to warn the user and rekey rather than failing here.

Reply:
I think including this as a requirement of having a VDS will make our lives easier. What are the cases where it won't have the alleles key? The combiner? It should be easy for us to return a keyed result from each step.

Reply:
actually, the combiner methods are broken if the alleles are part of the key. This is another reason to have all VDSes be keyed by both locus and alleles.